openworld/sdk/generator/client/visitors/models.py

```diff
@@ -101,263 +89,113 @@ def parse_datamodels(parser: OpenAPIParser) -> collections.defaultdict[str, Data
-    r"""Finds if parent has type attribute, removes all fields from it, and replaces them with one `__root__` attribute.
-
-    Args:
-        parent(DataModel): The parent model.
-        children(list[DataModel]): Children of a given parent model.
-
-    Returns:
-        DataModel: Refactored parent model.
     """
-    type_field = [field for field in parent.fields if field.name == "type"]
-    if not type_field:
-        return parent
-
-    type_field = type_field[0]
-
-    parent.fields.clear()
-
-    type_field.name = "__root__"
-
-    data_types = [DataType(type=child.class_name, reference=child.reference) for child in children]
-
-    type_field.data_type = DataType(data_types=data_types)
-
-    type_field.default = "Field(None, discriminator='type')"
-    type_field.required = True
-    type_field.use_field_description = False
+    children = list(filter(lambda model: model.base_class == parent_classname, models.values()))
+    return [child.class_name for child in children]
 
-    parent.fields.append(type_field)
-    return parent
 
-
-def has_type_attribute(model: DataModel) -> bool:
-    r"""Checks if a model has a `type` attribute.
+def parse_raw_discriminator(owner: str, raw_discriminator: dict[str, str], models: dict[str, DataModel]) -> Discriminator:
+    r"""Parses discriminator object from its raw representation parsed from the spec.
 
     Args:
-        model(DataModel): Model to be checked.
+        owner(str): Owner model of the discriminator, aka parent.
+        raw_discriminator(dict[str, str]): Raw discriminator object.
+        models(dict[str, DataModel]): A mapping for models, and their classnames as keys.
 
     Returns:
-        bool: rue if model has a `type` attribute, False otherwise.
-    """
-    for field in model.fields:
-        if field.name == "type":
-            return True
-    return False
-
-
-def is_parent_processed(parent: DataModel, children: list[DataModel]) -> bool:
-    """Checks if a parent model has been post processed before, in other words, it has children models, and one `__root__` field.
-
-    Args:
-        parent(DataModel): A model that is a parent for a set of children models by inheritance.
-        children(list[DataModel]): Children of a given parent model.
 
-    Returns:
-        bool: True if parent is processed, False otherwise.
     """
+    discriminator: Discriminator = parse_obj_as(Discriminator, {**raw_discriminator, "owner": owner})
 
-    return (
-        len(parent.fields) == 1
-        and (parent.fields[0].name == "__root__" or isinstance(parent, CustomRootType))
-        and len([child.class_name for child in children if child.class_name in parent.fields[0].type_hint])
-    )
+    if not discriminator.mapping:
+        discriminator.mapping = {
+            child_classname: child_classname for child_classname in parse_children_classnames(parent_classname=discriminator.owner, models=models)
```

Comment:
Why do we map the `child_classname` to itself? Is this a typo or intended logic?

Reply:
It is intentional, as the `mapping` section in OpenAPI specs is optional, and in case it's not used, we want to parse them into the `discriminator` object early to separate parsing & actually using `discriminator` objects.

For more info about the usage of mapping in the discriminator object, and what the default value of the discriminator property can be, please read the documentation [here](https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/#:~:text=discriminator%20property%20value.-,Mapping%20Type%20Names,-It%20is%20implied).